// ngx-translate-messageformat-compiler/esm2015/ngx-translate-messageformat-compiler.js
import { InjectionToken, Inject, Optional } from '@angular/core';
import { TranslateCompiler } from '@ngx-translate/core';
import * as MessageFormat from 'messageformat';
import { __rest } from 'tslib';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const MESSAGE_FORMAT_CONFIG = new InjectionToken("MESSAGE_FORMAT_CONFIG");
/**
 * @record
 */

const defaultConfig = {
    biDiSupport: false,
    formatters: undefined,
    locales: undefined,
    strictNumberSign: false
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * This compiler expects ICU syntax and compiles the expressions with messageformat.js
 */
class TranslateMessageFormatCompiler extends TranslateCompiler {
    /**
     * @param {?=} config
     */
    constructor(config) {
        super();
        config = Object.assign({}, defaultConfig, config);
        this.messageFormat = new MessageFormat(config.locales)
            .addFormatters(config.formatters)
            .setBiDiSupport(config.biDiSupport)
            .setStrictNumberSign(config.strictNumberSign);
    }
    /**
     * @param {?} value
     * @param {?} lang
     * @return {?}
     */
    compile(value, lang) {
        return this.messageFormat.compile(value, lang);
    }
    /**
     * @param {?} translations
     * @param {?} lang
     * @return {?}
     */
    compileTranslations(translations, lang) {
        return this.messageFormat.compile(translations, lang);
    }
}
/** @nocollapse */
TranslateMessageFormatCompiler.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MESSAGE_FORMAT_CONFIG,] },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/* tslint:disable-next-line no-console */
const log = (...message) => console.log(tag, ...message);
const tag = "[TranslateMessageFormatCompiler]";
class TranslateMessageFormatDebugCompiler extends TranslateMessageFormatCompiler {
    /**
     * @param {?} value
     * @param {?} lang
     * @return {?}
     */
    compile(value, lang) {
        log(`COMPILE (${lang})`, value);
        const /** @type {?} */ interpolationFn = super.compile(value, lang);
        return this.wrap(interpolationFn, value);
    }
    /**
     * @param {?} value
     * @param {?} lang
     * @return {?}
     */
    compileTranslations(value, lang) {
        log(`COMPILE (${lang})`, value);
        const _a = super.compileTranslations(value, lang), { toString } = _a, interpolationFns = __rest(_a, ["toString"]);
        return Object.assign({ toString }, this.wrapRecursively(interpolationFns, value));
    }
    /**
     * @param {?} fn
     * @param {?} reference
     * @return {?}
     */
    wrap(fn, reference) {
        return (params) => {
            log("INTERPOLATE", reference, params);
            return fn(params);
        };
    }
    /**
     * @param {?} obj
     * @param {?} referenceObj
     * @return {?}
     */
    wrapRecursively(obj, referenceObj) {
        return Object.keys(obj).reduce((acc, key) => {
            const /** @type {?} */ value = obj[key];
            const /** @type {?} */ referenceValue = referenceObj[key];
            return typeof value === "function"
                ? Object.assign({}, acc, { [key]: this.wrap(value, referenceValue) }) : Object.assign({}, acc, { [key]: this.wrapRecursively(value, referenceValue) });
        }, {});
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { TranslateMessageFormatCompiler, TranslateMessageFormatDebugCompiler, MESSAGE_FORMAT_CONFIG };
