// ngx-translate-messageformat-compiler/bundles/ngx-translate-messageformat-compiler.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-translate/core'), require('messageformat')) :
	typeof define === 'function' && define.amd ? define('ngx-translate-messageformat-compiler', ['exports', '@angular/core', '@ngx-translate/core', 'messageformat'], factory) :
	(factory((global['ngx-translate-messageformat-compiler'] = {}),global.ng.core,global.core$1,global.messageformat));
}(this, (function (exports,core,core$1,MessageFormat) { 'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0
THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.
See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */
var extendStatics = Object.setPrototypeOf ||
    ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
    function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
}







function __read(o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
}
function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++)
        ar = ar.concat(__read(arguments[i]));
    return ar;
}

var MESSAGE_FORMAT_CONFIG = new core.InjectionToken("MESSAGE_FORMAT_CONFIG");
var defaultConfig = {
    biDiSupport: false,
    formatters: undefined,
    locales: undefined,
    strictNumberSign: false
};
var TranslateMessageFormatCompiler = /** @class */ (function (_super) {
    __extends(TranslateMessageFormatCompiler, _super);
    function TranslateMessageFormatCompiler(config) {
        var _this = _super.call(this) || this;
        config = Object.assign({}, defaultConfig, config);
        _this.messageFormat = new MessageFormat(config.locales)
            .addFormatters(config.formatters)
            .setBiDiSupport(config.biDiSupport)
            .setStrictNumberSign(config.strictNumberSign);
        return _this;
    }
    TranslateMessageFormatCompiler.prototype.compile = function (value, lang) {
        return this.messageFormat.compile(value, lang);
    };
    TranslateMessageFormatCompiler.prototype.compileTranslations = function (translations, lang) {
        return this.messageFormat.compile(translations, lang);
    };
    return TranslateMessageFormatCompiler;
}(core$1.TranslateCompiler));
TranslateMessageFormatCompiler.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MESSAGE_FORMAT_CONFIG,] },] },
]; };
var log = function () {
    var message = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        message[_i] = arguments[_i];
    }
    return console.log.apply(console, __spread([tag], message));
};
var tag = "[TranslateMessageFormatCompiler]";
var TranslateMessageFormatDebugCompiler = /** @class */ (function (_super) {
    __extends(TranslateMessageFormatDebugCompiler, _super);
    function TranslateMessageFormatDebugCompiler() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TranslateMessageFormatDebugCompiler.prototype.compile = function (value, lang) {
        log("COMPILE (" + lang + ")", value);
        var interpolationFn = _super.prototype.compile.call(this, value, lang);
        return this.wrap(interpolationFn, value);
    };
    TranslateMessageFormatDebugCompiler.prototype.compileTranslations = function (value, lang) {
        log("COMPILE (" + lang + ")", value);
        var _a = _super.prototype.compileTranslations.call(this, value, lang), toString = _a.toString, interpolationFns = __rest(_a, ["toString"]);
        return Object.assign({ toString: toString }, this.wrapRecursively(interpolationFns, value));
    };
    TranslateMessageFormatDebugCompiler.prototype.wrap = function (fn, reference) {
        return function (params) {
            log("INTERPOLATE", reference, params);
            return fn(params);
        };
    };
    TranslateMessageFormatDebugCompiler.prototype.wrapRecursively = function (obj, referenceObj) {
        var _this = this;
        return Object.keys(obj).reduce(function (acc, key) {
            var value = obj[key];
            var referenceValue = referenceObj[key];
            return typeof value === "function"
                ? Object.assign({}, acc, (_b = {}, _b[key] = _this.wrap(value, referenceValue), _b)) : Object.assign({}, acc, (_c = {}, _c[key] = _this.wrapRecursively(value, referenceValue), _c));
            var _b, _c;
        }, {});
    };
    return TranslateMessageFormatDebugCompiler;
}(TranslateMessageFormatCompiler));

exports.TranslateMessageFormatCompiler = TranslateMessageFormatCompiler;
exports.TranslateMessageFormatDebugCompiler = TranslateMessageFormatDebugCompiler;
exports.MESSAGE_FORMAT_CONFIG = MESSAGE_FORMAT_CONFIG;

Object.defineProperty(exports, '__esModule', { value: true });

})));
