// ngx-translate-messageformat-compiler/esm5/ngx-translate-messageformat-compiler.js
import { __rest, __extends, __spread } from 'tslib';
import { InjectionToken, Inject, Optional } from '@angular/core';
import { TranslateCompiler } from '@ngx-translate/core';
import * as MessageFormat from 'messageformat';

var MESSAGE_FORMAT_CONFIG = new InjectionToken("MESSAGE_FORMAT_CONFIG");
var defaultConfig = {
    biDiSupport: false,
    formatters: undefined,
    locales: undefined,
    strictNumberSign: false
};
var TranslateMessageFormatCompiler = /** @class */ (function (_super) {
    __extends(TranslateMessageFormatCompiler, _super);
    function TranslateMessageFormatCompiler(config) {
        var _this = _super.call(this) || this;
        config = Object.assign({}, defaultConfig, config);
        _this.messageFormat = new MessageFormat(config.locales)
            .addFormatters(config.formatters)
            .setBiDiSupport(config.biDiSupport)
            .setStrictNumberSign(config.strictNumberSign);
        return _this;
    }
    TranslateMessageFormatCompiler.prototype.compile = function (value, lang) {
        return this.messageFormat.compile(value, lang);
    };
    TranslateMessageFormatCompiler.prototype.compileTranslations = function (translations, lang) {
        return this.messageFormat.compile(translations, lang);
    };
    return TranslateMessageFormatCompiler;
}(TranslateCompiler));
TranslateMessageFormatCompiler.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MESSAGE_FORMAT_CONFIG,] },] },
]; };
var log = function () {
    var message = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        message[_i] = arguments[_i];
    }
    return console.log.apply(console, __spread([tag], message));
};
var tag = "[TranslateMessageFormatCompiler]";
var TranslateMessageFormatDebugCompiler = /** @class */ (function (_super) {
    __extends(TranslateMessageFormatDebugCompiler, _super);
    function TranslateMessageFormatDebugCompiler() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TranslateMessageFormatDebugCompiler.prototype.compile = function (value, lang) {
        log("COMPILE (" + lang + ")", value);
        var interpolationFn = _super.prototype.compile.call(this, value, lang);
        return this.wrap(interpolationFn, value);
    };
    TranslateMessageFormatDebugCompiler.prototype.compileTranslations = function (value, lang) {
        log("COMPILE (" + lang + ")", value);
        var _a = _super.prototype.compileTranslations.call(this, value, lang), toString = _a.toString, interpolationFns = __rest(_a, ["toString"]);
        return Object.assign({ toString: toString }, this.wrapRecursively(interpolationFns, value));
    };
    TranslateMessageFormatDebugCompiler.prototype.wrap = function (fn, reference) {
        return function (params) {
            log("INTERPOLATE", reference, params);
            return fn(params);
        };
    };
    TranslateMessageFormatDebugCompiler.prototype.wrapRecursively = function (obj, referenceObj) {
        var _this = this;
        return Object.keys(obj).reduce(function (acc, key) {
            var value = obj[key];
            var referenceValue = referenceObj[key];
            return typeof value === "function"
                ? Object.assign({}, acc, (_b = {}, _b[key] = _this.wrap(value, referenceValue), _b)) : Object.assign({}, acc, (_c = {}, _c[key] = _this.wrapRecursively(value, referenceValue), _c));
            var _b, _c;
        }, {});
    };
    return TranslateMessageFormatDebugCompiler;
}(TranslateMessageFormatCompiler));

export { TranslateMessageFormatCompiler, TranslateMessageFormatDebugCompiler, MESSAGE_FORMAT_CONFIG };
